Multiprocess build support

ryanofsky opened the pull request:
This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/projects/10).

This splits autotools, depends build, and travis changes out of #10102, so code changes and build system changes can be reviewed separately.

ryanofsky committed:
Add multiprocess travis configuration

ryanofsky committed:
Set LD_LIBRARY_PATH consistently in travis tests

Remove inconsistency between functional and unit test environments and make it
possible to substitute bitcoin-qt and bitcoin-node in place of bitcoind in
python tests, or to link bitcoind against shared libraries.

ryanofsky committed:
libmultiprocess depends build

ryanofsky committed:
Multiprocess build changes

autotools and automake changes to support multiprocess execution.

This adds a new --enable-multiprocess flag, and build configuration code to
detect libraries needed for multiprocess support. The --enable-multiprocess
flag builds new bitcoin-node and bitcoin-gui executables, which are updated in
https://github.com/bitcoin/bitcoin/pull/10102 to communicate across processes.
But for now they are functionally equivalent to existing bitcoind and
bitcoin-qt executables.

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -143,8 +143,8 @@ def main(self):
-        self.options.bitcoind = os.getenv("BITCOIND", default=config["environment"]["BUILDDIR"] + '/src/bitcoind' + config["environment"]["EXEEXT"])
-        self.options.bitcoincli = os.getenv("BITCOINCLI", default=config["environment"]["BUILDDIR"] + '/src/bitcoin-cli' + config["environment"]["EXEEXT"])
+        self.options.bitcoind = os.path.join(config["environment"]["BUILDDIR"], os.getenv("BITCOIND", default='src/bitcoind' + config["environment"]["EXEEXT"]))
+        self.options.bitcoincli = os.path.join(config["environment"]["BUILDDIR"], os.getenv("BITCOINCLI", default='src/bitcoin-cli' + config["environment"]["EXEEXT"]))
```
Why is this changed?

We already export the PATH two lines below, so setting `BITCOIND=bitcoin-node` should be sufficient

ryanofsky replied:
Oh, I didn't know about the PATH feature, so I'll revert this.

I changed this to get `BITCOIND=src/bitcoin-node test/functional/test_runner.py` not to fail, but I should probably take advantage of the PATH setting and change `BITCOIND=src/bitcoin-node` to `BITCOIND=bitcoin-node` instead.

Sjors commented:
Adding `capnp` probably deserves it's own PR, but at least a separate commit.

Sjors commented on depends/packages/boost.mk:
```diff
@@ -1,7 +1,7 @@
 $(package)_version=1_70_0
 $(package)_download_path=https://dl.bintray.com/boostorg/release/1.70.0/source/
-$(package)_file_name=$(package)_$($(package)_version).tar.bz2
+$(package)_file_name=boost_$($(package)_version).tar.bz2
```
Is this an unrelated improvement?

ryanofsky replied:
It is related. I do think it's an improvement to not use our internal package name in a tarball URL that's generated and hosted by a third party and we don't control. But this change is actually needed to make the build work.

Specifically, this is needed to make native_boost package work, since it derives from the boost package but has a different package name. The native_boost package is a new package needed to build the native_libmultiprocess package which currently uses boost::optional. The native_libmultiprocess package is needed to run a code generation tool in cross-compiled builds (since a tool that runs at build time should be native, not cross-compiled). I'm probably going to try to drop the boost::optional dependency at some point, but for now just adding a native_boost package seemed like the path of least resistance, since boost::optional is so convenient.

Sjors replied:
Now that `native_libmultiprocess` doesn't need `boost::optional`, do you still need this? If you do need it, maybe make that a separate commit.

ryanofsky commented:
> On macOS, I looks like I need to do more than just brew install cmake. See flood of errors.

Thanks for the log! It looks like the error is:

```
#error "This code requires C++14. Either your compiler does not support it or it is not enabled."
```

So probably what's happening is that for whatever reason my compiler defaults to c++14, while your compiler defaults to an older c++ version. It looks like I could fix the libmultiprocess build to specify this:

https://cmake.org/cmake/help/latest/command/target_compile_features.html
https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES

An alternate approach might be to downgrade capnp from 0.7.0, since 0.7.0 is the first version that requires c++14, and I'm actually not relying on any 0.7.0 features.

I do need to try building and running this code again on mac soon, since I haven't tried it in a while.

Sjors commented:
See also #13356 for C++14 discussion. However, as long as multiprocess is optional, I think it's more future proof to use the C++14, with the latest `capnp` version, and fix the macOS build.

ariard reviewed (commented):
OK, I succeed to build `bitcoin-gui` and `bitcoin-node` and `bitcoin-wallet` on Debian 9. 

Main problem I had to build first libmultiprocess as a dependency and then pass its library directory with `--with-libmultiprocess` to `./configure` which isn't documented in the `multiprocess.md`. Then to run `bitcoin-node` and `bitcoin-wallet` I had to add `libcapnp-rpc-0.7.0.so` in runtime search path. But it succeed and both are working successfully. Don't try to run bitcoin-gui as I was lacking more dependencies.

IMO you should update `doc/build-unix.md` with a multiprocess section which tell you exactly what to do step by step and requirements. And we should blindly follow these sections on Debian/Fedora/Gentoo and see if it's smooth enough for a user.

But overall, super excited to go forward with multiprocess !

ryanofsky reviewed (commented):
Thanks for testing, ariard! It sounds like you encountered bugs that you worked around, but I'd like to know more specifics about what wasn't working. I'm happy to help debug anytime, or if you want to work online, you can post issues to https://github.com/chaincodelabs/libmultiprocess/issues/new (general issues, bitcoin-specific issues, and support requests all welcome there). Regarding the specific workarounds:

* Passing a prefix directory to --with-libmultiprocess (or specifying it at all) shouldn't be necessary, and I actually removed the option to hardcode a prefix in some new changes I haven't pushed yet. The configure script should be able to locate dependencies through pkgconf. If that isn't happening please let me know the specifics or open an issue!

* Regarding documentation, `configure --help` text should serve as a reference for config options, and the `--enable-multiprocess` error messages should provide clear guidance if anything goes wrong. If the help text isn't helpful, I'd want to open an issue to track that. [multiprocess.md](https://github.com/ryanofsky/bitcoin/blob/pr/ipc-build/doc/multiprocess.md) and libmultiprocess [README.md](https://github.com/chaincodelabs/libmultiprocess/blob/master/README.md) could definitely be improved, but for install steps I would mostly want to add to the existing install steps rather than add separate instructions. So far I've only really looked at the depends instructions, not the unix/mac/windows instructions, and I need to get to those at some point.

* The problem with library paths sounds like a bug. Rpaths should be set correctly so you shouldn't need to anything special at runtime.

ariard commented:
For others reviewers, after talking with Russ, my mistake was not to use depends prefix as described in `depends/README.md`. Will give it another try after updates.

ariard commented on depends/README.md:
```diff
@@ -73,6 +73,7 @@ The following can be set when running make: make FOO=bar
     NO_UPNP: Don't download/build/cache packages needed for enabling upnp
     DEBUG: disable some optimizations and enable more runtime checking
     RAPIDCHECK: build rapidcheck (experimental, requires cmake)
+    MULTIPROCESS: build libmultiprocess (experimental, requires cmake)
```
Thanks for the great explication, bookmarked!

ryanofsky commented:
re: http://gnusha.org/bitcoin-builds/2019-08-16.log

> \<dongcarl> cfields: I think the boostrap part of `boost` is actually "native", as in, should be moved to a `boost_native` package like the native part of `protobuf` was... Does that ring somewhat true to you?

@dongcarl this PR adds a [native_boost](https://github.com/ryanofsky/bitcoin/blob/pr/ipc-build/depends/packages/native_boost.mk) package, but there should be no need for a native package just to use the bootstrap tool, unless you need to use the bootstrap tool from other packages, and not just use it internally in the boost package.

fanquake commented:
I depends built, compiled and played around with this (macOS). Going to re-read all the discussion in #13900 

I'm also going to extract https://github.com/bitcoin/bitcoin/pull/16367/commits/7149b1ae67f0e3508407f1b668b11ee8e7456ab3 and https://github.com/bitcoin/bitcoin/pull/16367/commits/a295b9decb14534a98f6c993df9a9379ce752aeb into a separate PR. They are clear improvements, and will likely clear up some confusion here and in other depends related PRs.

dongcarl commented on depends/packages/boost.mk:
```diff
@@ -1,7 +1,7 @@
 $(package)_version=1_70_0
 $(package)_download_path=https://dl.bintray.com/boostorg/release/1.70.0/source/
-$(package)_file_name=$(package)_$($(package)_version).tar.bz2
+$(package)_file_name=boost_$($(package)_version).tar.bz2
```
Perhaps it's cleaner to inherit from the `native` package like you did for `capnp` and `libmultiprocess`? Maybe I'm missing something here though.

ryanofsky replied:
> Perhaps it's cleaner to inherit from the `native` package like you did for `capnp` and `libmultiprocess`? Maybe I'm missing something here though.

That'd be another approach. I wouldn't mind adding that, but it would make the change to the boost package here a little bigger and less minimal than actually needed for the PR.

Also, I do think the change here is an improvement by itself. It seems janky to me tie our local package name to the convention the upstream project is using for download urls.

theuni replied:
+1. It's fine when it's convenient shorthand or more clear, but no need to keep them tied if it makes anything harder.

dongcarl commented on ci/test/06_script_b.sh:
```diff
@@ -16,13 +16,13 @@ fi
 if [ "$RUN_FUZZ_TESTS" = "true" ]; then
   BEGIN_FOLD fuzz-tests
-  DOCKER_EXEC test/fuzz/test_runner.py -l DEBUG ${DIR_FUZZ_IN}
+  DOCKER_EXEC LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib test/fuzz/test_runner.py -l DEBUG ${DIR_FUZZ_IN}
```
Curious: where is this used?

dongcarl commented on depends/funcs.mk:
```diff
@@ -130,6 +130,7 @@ $(1)_config_env+=$($(1)_config_env_$(host_arch)_$(host_os)) $($(1)_config_env_$(
 
 $(1)_config_env+=PKG_CONFIG_LIBDIR=$($($(1)_type)_prefix)/lib/pkgconfig
 $(1)_config_env+=PKG_CONFIG_PATH=$($($(1)_type)_prefix)/share/pkgconfig
+$(1)_config_env+=CMAKE_MODULE_PATH=$($($(1)_type)_prefix)/lib/cmake
```
Curious: which package recipe will write CMAKE modules to the prefix? Are they used by other depends packages or by the non-depends Bitcoin build?

dongcarl commented on src/Makefile.qt.include:
```diff
@@ -397,42 +402,56 @@ QT_FORMS_H=$(join $(dir $(QT_FORMS_UI)),$(addprefix ui_, $(notdir $(QT_FORMS_UI:
+bitcoin_qt_common_ldadd = qt/libbitcoinqt.a $(LIBBITCOIN_SERVER)
 if ENABLE_WALLET
-qt_bitcoin_qt_LDADD += $(LIBBITCOIN_UTIL) $(LIBBITCOIN_WALLET)
+bitcoin_qt_common_ldadd += $(LIBBITCOIN_UTIL) $(LIBBITCOIN_WALLET) $(LIBBITCOIN_SERVER)
```
The last `$(LIBBITCOIN_SERVER)` is probably unneeded...

hebasto replied:
Yeah, it seems unconditionally included in:
https://github.com/bitcoin/bitcoin/blob/d6921b176186fd8304b6d6a13772fd0fc32fe726/src/Makefile.qt.include#L313

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r340334676

> The last `$(LIBBITCOIN_SERVER)` is probably unneeded...

Removed with Hennadii's diff https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-595371624

dongcarl commented on src/Makefile.qt.include:
```diff
@@ -397,42 +402,56 @@ QT_FORMS_H=$(join $(dir $(QT_FORMS_UI)),$(addprefix ui_, $(notdir $(QT_FORMS_UI:
+bitcoin_gui_CPPFLAGS = $(bitcoin_qt_common_cppflags)
+bitcoin_gui_CXXFLAGS = $(bitcoin_qt_common_cxxflags)
+bitcoin_gui_SOURCES = $(bitcoin_qt_common_sources)
+bitcoin_gui_LDADD = $(bitcoin_qt_common_ldadd) $(LIBBITCOIN_COMMON)
```
The last `$(LIBBITCOIN_COMMON)` is probably unneeded...

ryanofsky replied:
> The last `$(LIBBITCOIN_COMMON)` is probably unneeded...

Will try to remove, but not so sure about this one. The util and wallet libraries do (no longer) depend on the server library in the previous line you commented on, but a bunch of the earlier libraries do depend on the common library, so it's not surprising that might need to be repeated.

hebasto replied:
> The last `$(LIBBITCOIN_COMMON)` is probably unneeded...

It seems unconditionally included in:
https://github.com/bitcoin/bitcoin/blob/d6921b176186fd8304b6d6a13772fd0fc32fe726/src/Makefile.qt.include#L320-L322

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r340337314

> The last `$(LIBBITCOIN_COMMON)` is probably unneeded...

Removed with Hennadii's diff https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-595371624

ryanofsky reviewed (commented):
Thanks for the review!

ryanofsky commented:
Rebased 6a84d0553f712a3e228fd697d3a831d21f210fa5 -> 71aa1a2c9be6ded8daccf6e8af318b82f3a0f0ca ([`pr/ipc-build.13`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.13) -> [`pr/ipc-build.14`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.13-rebase..pr/ipc-build.14)) due to conflict with #18166

Sjors commented:
I'm unable to build depends on macOS 10.15.3: `make MULTIPROCESS=1` results in `A C++ library with support for C++11 features is required`, see [log](https://gist.github.com/Sjors/7570fd3d67238b6e1dbdacaeb0a7fb20#file-make-depends-log).

Sjors commented:
Concept ACK. I was able to build with Ubuntu and depends. I'll try Ubuntu and libmultiprocess  install later, as well as macOS.

https://github.com/chaincodelabs/libmultiprocess/pull/25 lets us drop boost_native from this PR which is nice.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-592093081

> I'm unable to build depends on macOS 10.15.3: `make MULTIPROCESS=1` results in `A C++ library with support for C++11 features is required`, see [log](https://gist.github.com/Sjors/7570fd3d67238b6e1dbdacaeb0a7fb20#file-make-depends-log) (Apple clang version 11.0.0 (clang-1100.0.33.17))

The same "configure: error: *** A C++ library with support for C++11 features is required." was also reported https://github.com/chaincodelabs/libmultiprocess/issues/5, and the fix seemed to be to be fixed by some combination of upgrading and cleaning the build.

FWIW, I'm also able to reproduce the problem. Looking in `depends/work/build/x86_64-apple-darwin18.0.0/native_capnp/0.7.0-7b1af401b59/config.log` the detailed error is:

```
configure:5116: /Library/Developer/CommandLineTools/usr/bin/clang++ -std=gnu++14 -stdlib=libc++ -c -O2 -DNDEBUG -I/Users/user/src/bitcoin/depends/x86_64-apple-darwin18.0.0/native/include     conftest.cpp >&5
In file included from conftest.cpp:13:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/unordered_map:369:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__hash_table:16:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/memory:653:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/typeinfo:61:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/exception:82:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:86:
/Library/Developer/CommandLineTools/usr/include/c++/v1/stdlib.h:94:15: fatal error: 'stdlib.h' file not found
#include_next <stdlib.h>
              ^~~~~~~~~~
1 error generated.
```

which according to https://discourse.brew.sh/t/clang-can-no-longer-find-usr-include-header-files-fatal-error-stdlib-h-file-not-found/4523/6 has something to do with mac os removing /usr/include, but it's unclear how this should be fixed.

Just running checking out this PR and running "make -C depends NO_QT=1 MULTIPROCESS=1" on a mac was sufficient to reproduce this

ryanofsky commented:
I can reproduce the mac depends problem independently without bitcoin or depends:

```bash
$ echo '#include <stdlib.h>' > test.cpp
$ /Library/Developer/CommandLineTools/usr/bin/clang++ -c test.cpp
In file included from test.cpp:1:
/Library/Developer/CommandLineTools/usr/include/c++/v1/stdlib.h:94:15: fatal error: 'stdlib.h' file not found
#include_next <stdlib.h>
              ^~~~~~~~~~
1 error generated.
```

Adding `-isysroot` seems to fix it, so I guess it needs to be added to the depends build:

```bash
$ /Library/Developer/CommandLineTools/usr/bin/clang++ -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -c test.cpp
```

ryanofsky commented:
@Sjors, the following change gets past the reported mac os problem https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-592093081: 

```diff
diff --git a/depends/builders/darwin.mk b/depends/builders/darwin.mk
index eb587fca890..69c394ec1db 100644
--- a/depends/builders/darwin.mk
+++ b/depends/builders/darwin.mk
@@ -1,5 +1,5 @@
-build_darwin_CC:=$(shell xcrun -f clang)
-build_darwin_CXX:=$(shell xcrun -f clang++)
+build_darwin_CC:=$(shell xcrun -f clang) --sysroot $(shell xcrun --show-sdk-path)
+build_darwin_CXX:=$(shell xcrun -f clang++) --sysroot $(shell xcrun --show-sdk-path)
 build_darwin_AR:=$(shell xcrun -f ar)
 build_darwin_RANLIB:=$(shell xcrun -f ranlib)
 build_darwin_STRIP:=$(shell xcrun -f strip)
```

I'll add it to the PR assuming the rest of the build succeeds. Thanks for testing and reporting!

ryanofsky committed:
depends: Use default macos clang compiler

Suggested by Cory Fields <cory-nospam-@coryfields.com>
https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-595393546
as alternate workaround for problem described
https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-594600985

hebasto commented:
Concept ACK.

hebasto commented on configure.ac:
```diff
@@ -236,6 +236,24 @@ if test x$enable_bip70 != xno; then
+
+AC_ARG_WITH([mpgen],
+  [AS_HELP_STRING([--with-mpgen=yes|no|auto|PREFIX],
+  [Build with libmultiprocess codegen tool. Useful to specify different libmultiprocess host system library and build system codegen tool prefixes when cross-compiling (default is host system libmultiprocess prefix])],
```
nit: a missed paren in the end of help string:
```suggestion
  [Build with libmultiprocess codegen tool. Useful to specify different libmultiprocess host system library and build system codegen tool prefixes when cross-compiling (default is host system libmultiprocess prefix)])],
```

hebasto commented on configure.ac:
```diff
@@ -236,6 +236,24 @@ if test x$enable_bip70 != xno; then
 
+AC_ARG_WITH([libmultiprocess],
+  [AS_HELP_STRING([--with-libmultiprocess=yes|no|auto],
+  [Build with libmultiprocess library. (default: detect with pkg-config)])],
```
nit: The help string could clear state that default=auto:

"(default is auto, i.e., detect with pkg-config)"

or smth similar.

hebasto commented on depends/config.site.in:
```diff
@@ -25,6 +28,10 @@ if test -z $enable_wallet && test -n "@no_wallet@"; then
+if test -z $enable_multiprocess && test -n "@multiprocess@"; then
+  enable_multiprocess=yes
+fi
+
```
This is a very convenient feature.
But, OTOH, it is a bit confusing as `--enable-multiprocess` help string states that "default is **no**".

May I suggest to drop this addition?

As another approach, the default value of `--enable-multiprocess` could be `auto` (but not now, when it is an experimental feature).

theuni replied:
Good points. I think as long as this is experimental, probably best to play it safe and require it to be --enabled by hand. Arguably it already has been because of the way depends was built, but it's easy to forget when bouncing around branches.

Not a strong preference, though.

hebasto commented on src/Makefile.am:
```diff
@@ -515,25 +520,25 @@ libbitcoin_cli_a_SOURCES = \
-bitcoind_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
-bitcoind_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
-bitcoind_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
+# bitcoind, bitcoin-wallet, bitcoin-node binaries #
```
I think the comment in its current state is correct as `bitcoind_common_{cppflags|cxxflags|ldflags}` are used to define `bitcoin_wallet_{CPPFLAGS|CXXFLAGS|LDFLAGS}` as well.

hebasto commented:
From [Cap’n Proto docs](https://capnproto.org/install.html#supported-compilers):
> The minimum versions are:
>    - GCC 4.9
>    - Clang 3.5

[Our requirements](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) are:
- GCC 4.8+
- Clang 3.3+

Could this be an issue?

hebasto reviewed (approved):
ACK d6921b176186fd8304b6d6a13772fd0fc32fe726, tested on Linux Mint 19.3 (gcc 7.4.0) and Fedora 31 (gcc 9.2.1); built depends with `MULTIPROCESS=1`, and with/without `NO_QT=1`.

Verified new downloads (`capnp` and `libmultiprocess`) paths and hashes.

In commit "Add multiprocess travis configuration" (f077d9cb46c6b825ff560f7c8a1b655a8d9fc2be) could an environment file be used as it has done in #16623?

theuni commented on src/Makefile.am:
```diff
@@ -548,22 +552,22 @@ libbitcoin_cli_a_SOURCES = \
-bitcoind_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
-bitcoind_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
+# bitcoind, bitcoin-wallet, bitcoin-node binaries #
+bitcoind_common_sources = bitcoind.cpp
```
Please make these uppercase so they stand out from automake variables. As-is, this looks like a new helper lib.

ryanofsky replied:
> Please make these uppercase so they stand out from automake variables. As-is, this looks like a new helper lib.

I don't think this works. If I make everything uppercase I see errors:

```
src/Makefile.am:568: warning: variable 'BITCOIND_COMMON_LDFLAGS' is defined but no program or
src/Makefile.am:568: library has 'BITCOIND_COMMON' as canonical name (possible typo)
```

These are lowercase specifically because automake seems to pick up uppercase suffixes as target properties.

Tested with:

```bash
for f in bitcoind_common_sources bitcoind_common_cppflags bitcoind_common_cxxflags bitcoind_common_ldflags bitcoin_qt_common_cppflags bitcoin_qt_common_cxxflags bitcoin_qt_common_sources bitcoin_qt_common_ldadd bitcoin_qt_common_ldflags bitcoin_qt_common_libtoolflags; do
git grep -l $f | xargs sed -i "s/$f/$(echo $f | tr '[:lower:]' '[:upper:]')/g"
done
```

hebasto replied:
Agree that naming of convenient variables with `all_lower_case` makes them distinguishable from the Automake special variables.

theuni commented on src/Makefile.qt.include:
```diff
@@ -294,29 +299,43 @@ QT_FORMS_H=$(join $(dir $(QT_FORMS_UI)),$(addprefix ui_, $(notdir $(QT_FORMS_UI:
-# bitcoin-qt binary #
-qt_bitcoin_qt_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) $(BITCOIN_QT_INCLUDES) \
+# bitcoin-qt and bitcoin-gui binaries #
+bitcoin_qt_common_cppflags = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) $(BITCOIN_QT_INCLUDES) \
```
Same with these, please use uppercase for helper vars.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r388470002

> Same with these, please use uppercase for helper vars.

Seeing

```
src/Makefile.qt.include:323: warning: variable 'BITCOIN_QT_COMMON_LDFLAGS' is defined but no program or
src/Makefile.qt.include:323: library has 'BITCOIN_QT_COMMON' as canonical name (possible typo)
```

with this

hebasto commented:
FWIW, with applied diff:
```diff
diff --git a/src/Makefile.qt.include b/src/Makefile.qt.include
index c5da67b0e..7924917d1 100644
--- a/src/Makefile.qt.include
+++ b/src/Makefile.qt.include
@@ -312,7 +312,7 @@ if TARGET_WINDOWS
 endif
 bitcoin_qt_common_ldadd = qt/libbitcoinqt.a $(LIBBITCOIN_SERVER)
 if ENABLE_WALLET
-bitcoin_qt_common_ldadd += $(LIBBITCOIN_UTIL) $(LIBBITCOIN_WALLET) $(LIBBITCOIN_SERVER)
+bitcoin_qt_common_ldadd += $(LIBBITCOIN_UTIL) $(LIBBITCOIN_WALLET)
 endif
 if ENABLE_ZMQ
 bitcoin_qt_common_ldadd += $(LIBBITCOIN_ZMQ) $(ZMQ_LIBS)
@@ -333,7 +333,7 @@ qt_bitcoin_qt_LIBTOOLFLAGS = $(bitcoin_qt_common_libtoolflags)
 bitcoin_gui_CPPFLAGS = $(bitcoin_qt_common_cppflags)
 bitcoin_gui_CXXFLAGS = $(bitcoin_qt_common_cxxflags)
 bitcoin_gui_SOURCES = $(bitcoin_qt_common_sources)
-bitcoin_gui_LDADD = $(bitcoin_qt_common_ldadd) $(LIBBITCOIN_COMMON)
+bitcoin_gui_LDADD = $(bitcoin_qt_common_ldadd)
 bitcoin_gui_LDFLAGS = $(bitcoin_qt_common_ldflags)
 bitcoin_gui_LIBTOOLFLAGS = $(bitcoin_qt_common_libtoolflags)
```

I have built:
```
$ ./autogen.sh
$ make -C depends MULTIPROCESS=1
$ CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site ./configure
$ make
```

and then successfully run all Qt binaries:
```
$ ./src/qt/test/test_bitcoin-qt
```
```
$ ./src/qt/bitcoin-qt
```
```
$ ./src/bitcoin-gui
```

theuni commented on depends/packages/native_capnp.mk:
```diff
@@ -0,0 +1,17 @@
+package=native_capnp
+$(package)_version=0.7.0
+$(package)_download_path=https://capnproto.org/
+$(package)_file_name=capnproto-c++-$($(package)_version).tar.gz
```
IIRC the "+" in the filename gives some tools/filesystems grief, which is why we hack it out in native_cctools.
You can use (untested):
```
$(package)_download_file=capnproto-c++-$($(package)_version).tar.gz
$(package)_file_name=capnproto-c++-$($(package)_version).tar.gz
```
I'll try to track down a good explanation for why we do that first, though. Maybe it's no longer necessary.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r388486877

> IIRC the "+" in the filename gives some tools/filesystems grief, which is why we hack it out in native_cctools.

Thanks. It'd be nice to not have to do this if you find out it's unnecessary, but replaced + with x for now

theuni reviewed (commented):
Concept ACK to the build/depends changes. I've been absent for the libmultiprocess conversation thus far, so I'm just reviewing with the assumption that it's sane.

Very nice work on the depends changes. I left a few nits, but for the most part that all looks good from a quick glance.

theuni commented:
> @Sjors, the following change gets past the reported mac os problem [#16367 (comment)](https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-592093081):
> 
> ```diff
> diff --git a/depends/builders/darwin.mk b/depends/builders/darwin.mk
> index eb587fca890..69c394ec1db 100644
> --- a/depends/builders/darwin.mk
> +++ b/depends/builders/darwin.mk
> @@ -1,5 +1,5 @@
> -build_darwin_CC:=$(shell xcrun -f clang)
> -build_darwin_CXX:=$(shell xcrun -f clang++)
> +build_darwin_CC:=$(shell xcrun -f clang) --sysroot $(shell xcrun --show-sdk-path)
> +build_darwin_CXX:=$(shell xcrun -f clang++) --sysroot $(shell xcrun --show-sdk-path)
>  build_darwin_AR:=$(shell xcrun -f ar)
>  build_darwin_RANLIB:=$(shell xcrun -f ranlib)
>  build_darwin_STRIP:=$(shell xcrun -f strip)
> ```
> 
> I'll add it to the PR assuming the rest of the build succeeds. Thanks for testing and reporting!

These tools have changed significantly since this was created. Might be easier (and more correct these days?) to just switch to using "clang" and "clang++" here. I'm pretty sure that's essentially what we're recreating anyway.

ryanofsky reviewed (commented):
Thanks for the testing and review!

Updated d6921b176186fd8304b6d6a13772fd0fc32fe726 -> 4968ddfd9e9b570503edef7707ca24ad4bee87ab ([`pr/ipc-build.16`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.16) -> [`pr/ipc-build.17`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.16..pr/ipc-build.17)) with suggestions

re: https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-595349319

> The minimum versions are [...] Could this be an issue?

I don't see an issue here, but let me know if you have a specific suggestion or want to describe a problem in more detail. I think each project is describing its build requirements accurately right now. It's also possible to install capnproto without installing any compiler, for example by using an os package (https://packages.ubuntu.com/bionic/capnproto). We can write more documentation about this, but I think we should not add documentation about building or installing other projects if there isn't a specific reason to think it will be helpful to someone.

re: https://github.com/bitcoin/bitcoin/pull/16367#pullrequestreview-369782143

> In commit "Add multiprocess travis configuration" ([f077d9c](https://github.com/bitcoin/bitcoin/commit/f077d9cb46c6b825ff560f7c8a1b655a8d9fc2be)) could an environment file be used as it has done in #16623?

Thanks, done. It was just like this because this PR preceded #16623.

re: https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-595371624

> FWIW, with applied diff:

Thanks, applied diff

hebasto commented on configure.ac:
```diff
@@ -236,6 +236,24 @@ if test x$enable_bip70 != xno; then
 
+AC_ARG_WITH([libmultiprocess],
+  [AS_HELP_STRING([--with-libmultiprocess=yes|no|auto],
+  [Build with libmultiprocess library. (default: detect is auto, i.e. detect with pkg-config)])],
```
It seems now the first "detect" is unneeded, no?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r388529793

> It seems now the first "detect" is unneeded, no?

Thanks, fixed

ryanofsky reviewed (commented):
Updated 4968ddfd9e9b570503edef7707ca24ad4bee87ab -> 04bec8d2844e0cd2e09a62b21b006292f33f8093 ([`pr/ipc-build.17`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.17) -> [`pr/ipc-build.18`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.17..pr/ipc-build.18)) with new suggestions from Hennadii and cory. Replaced mac os workaround 4968ddfd9e9b570503edef7707ca24ad4bee87ab with 04bec8d2844e0cd2e09a62b21b006292f33f8093

Sjors commented:
I can build 04bec8d2844e0cd2e09a62b21b006292f33f8093 on macOS now, with `capnp` 0.7.0 via Homebrew and [libmultiprocess](https://github.com/chaincodelabs/libmultiprocess) installed with its latest macOS fix. I'll try again once the boost obliteration from https://github.com/chaincodelabs/libmultiprocess/pull/25 is used here.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-595880087

> I can build [04bec8d](https://github.com/bitcoin/bitcoin/commit/04bec8d2844e0cd2e09a62b21b006292f33f8093) on macOS now, with `capnp` 0.7.0 via Homebrew and [libmultiprocess](https://github.com/chaincodelabs/libmultiprocess) installed with its latest macOS fix. I'll try again once the boost obliteration from [chaincodelabs/libmultiprocess#25](https://github.com/chaincodelabs/libmultiprocess/pull/25) is used here.

Note: boost PR is already incorporated here:

*  https://github.com/bitcoin/bitcoin/blob/04bec8d2844e0cd2e09a62b21b006292f33f8093/depends/packages/native_libmultiprocess.mk#L2
* https://github.com/chaincodelabs/libmultiprocess/commit/c0e3a5045596872cb63110cba0ef17abae56fd07

hebasto reviewed (approved):
re-ACK 04bec8d2844e0cd2e09a62b21b006292f33f8093, only suggested changes applied since d6921b176186fd8304b6d6a13772fd0fc32fe726 (https://github.com/bitcoin/bitcoin/pull/16367#pullrequestreview-369782143).

Sjors commented:
Copying my earlier inline comment, since I can hardly find it:

Now that `native_libmultiprocess` doesn't need `boost::optional`, do you still need to add `native_boost`? If you do need it, maybe make that a separate commit.

ryanofsky commented:
> Copying my earlier inline comment, since I can hardly find it:
> 
> Now that `native_libmultiprocess` doesn't need `boost::optional`, do you still need to add `native_boost`? If you do need it, maybe make that a separate commit.

The three main commits in the is PR are adding (1) autotools (2) travis and (3) depends support for libmultiprocess, and there are also two smaller commits with travis and depends fixes that are needed multiprocess support but make sense independently. I'm pretty sure breaking up the depends commit fab9fe51b999cca7e9d2122f9baa1786b77f4cb0 would just make it harder to understand because it's not that big right now and split up it'd be harder to see that packages there are following consistent conventions and see where new definitions are referenced.

As for removing native_boost, I think it would work temporarily, but I don't want to remove it because https://github.com/chaincodelabs/libmultiprocess/pull/25 isn't the end of the story on boost support. I merged that PR because I like the way it removed extraneous boost usages and dropped the boost requirement. But I'm pretty sure I want to add back some boost support. At least an optional `libmultiprocess` library dependency on boost to drop the [duplicative boost::optional code I had to add](https://github.com/ryanofsky/bitcoin/blob/f229ddea365c89a173c4dc8870d445e78716077e/src/interfaces/capnp/common-types.h#L238-L278) in #10102 and an `mpgen` dependency on boost to get rid of verbose and [nonportable fork exec stuff](https://github.com/chaincodelabs/libmultiprocess/blob/c0e3a5045596872cb63110cba0ef17abae56fd07/src/mp/gen.cpp#L141-L152) there right now.

ryanofsky commented:
Updated 04bec8d2844e0cd2e09a62b21b006292f33f8093 -> 1f9a2699a3e4b6ccdf4912600ccfdbf0c8eb20a4 ([`pr/ipc-build.18`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.18) -> [`pr/ipc-build.19`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.18..pr/ipc-build.19)) just switching to `$DEPENDS_DIR` instead of using `$TRAVIS_BUILD_DIR/depends`

hebasto commented:
re-ACK ac0f2fb1095d40b3a851150f38199fc2966b546f, since previous [review](https://github.com/bitcoin/bitcoin/pull/16367#pullrequestreview-370794394):
- `$TRAVIS_BUILD_DIR/depends` --> `$DEPENDS_DIR`
- `libmultiprocess` updated to https://github.com/chaincodelabs/libmultiprocess/tree/78f2f75d4722c42988988efd2b0074d3ec64099e

MarcoFalke commented on src/Makefile.am:
```diff
@@ -572,7 +576,19 @@ bitcoind_LDADD = \
   $(LIBSECP256K1)
 
-bitcoind_LDADD += $(BOOST_LIBS) $(BDB_LIBS) $(MINIUPNPC_LIBS) $(EVENT_PTHREADS_LIBS) $(EVENT_LIBS) $(ZMQ_LIBS)
+bitcoind_common_ldadd += $(BOOST_LIBS) $(BDB_LIBS) $(MINIUPNPC_LIBS) $(EVENT_PTHREADS_LIBS) $(EVENT_LIBS) $(ZMQ_LIBS) $(LIBMULTIPROCESS_LIBS)
```
Why is LIBMULTIPROCESS_LIBS neeeded for the single-binary `bitcoind`?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r396726227

> Why is LIBMULTIPROCESS_LIBS neeeded for the single-binary `bitcoind`?

It's not needed, removed entirely from this PR and removed from bitcoind link line in #10102

MarcoFalke commented on src/Makefile.am:
```diff
@@ -616,29 +632,14 @@ bitcoin_tx_LDADD += $(BOOST_LIBS)
-bitcoin_wallet_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
+bitcoin_wallet_CPPFLAGS = $(bitcoind_common_cppflags)
+bitcoin_wallet_CXXFLAGS = $(bitcoind_common_cxxflags)
+bitcoin_wallet_LDFLAGS = $(bitcoind_common_ldflags)
```
Why are the bitcoind common flags used for the bitcoin wallet tool? Could remove the `d`?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16367#discussion_r396729362

> Why are the bitcoind common flags used for the bitcoin wallet tool? Could remove the `d`?

Makes sense, changed the names as suggested

maflcko reviewed (commented):
Some dumb style questions

ryanofsky reviewed (commented):
Thanks for reviews!

Updated ac0f2fb1095d40b3a851150f38199fc2966b546f -> b53cf5a3f73ecdd57e4b421f5cb26bfe175868a6 ([`pr/ipc-build.20`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.20) -> [`pr/ipc-build.21`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.20..pr/ipc-build.21)) with suggested makefile style cleanups, also bumped libmultiprocess depends version again

ryanofsky commented:
Updated b53cf5a3f73ecdd57e4b421f5cb26bfe175868a6 -> 06d2f9128a4f5144cc459e04b7b6d2ab3316b105 ([`pr/ipc-build.21`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.21) -> [`pr/ipc-build.22`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.21..pr/ipc-build.22)) to fix link errors building `bitcoin-wallet`:

```
libbitcoin_util.a(libbitcoin_util_a-request.o): In function `JSONRPCReply[abi:cxx11](UniValue const&, UniValue const&, UniValue const&)':
src/rpc/request.cpp:48: undefined reference to `UniValue::write[abi:cxx11](unsigned int, unsigned int) const'
libbitcoin_util.a(libbitcoin_util_a-system.o): In function `ArgsManager::logArgsPrefix(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<UniValue, std::allocator<UniValue> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<UniValue, std::allocator<UniValue> > > > > const&) const':
src/util/system.cpp:875: undefined reference to `UniValue::write[abi:cxx11](unsigned int, unsigned int) const'
```

hebasto commented:
Tested 06d2f9128a4f5144cc459e04b7b6d2ab3316b105 on Linux Mint 19.3 with depends.

Yes, it builds successfully, but...

> > Why is LIBMULTIPROCESS_LIBS neeeded for the single-binary `bitcoind`?
>
> It's not needed, removed entirely from this PR

It seems questionable for me not using `LIBMULTIPROCESS_LIBS` and `LIBMULTIPROCESS_CFLAGS` in `src/Makefile.am` entirely, as `Libs` and `Cflags` in `lib/pkgconfig/libmultiprocess.pc` are not empty.

ryanofsky commented:
Rebased 06d2f9128a4f5144cc459e04b7b6d2ab3316b105 -> b919efadff3d0393f4a8c3c1dc735f7ac5c665bb ([`pr/ipc-build.22`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.22) -> [`pr/ipc-build.23`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build.22-rebase..pr/ipc-build.23)) due to conflict with #18514

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-605514556

> It seems questionable for me not using `LIBMULTIPROCESS_LIBS`

The point of this PR is to reduce the size and review burden of #10102. In order for that to be possible and to not do everything in a single giant PR, I need to be able to make smaller PRs (listed [here](https://github.com/bitcoin/bitcoin/projects/10)) that add changes bits of functionality which are initially unused. This PR pulls out all the more complicated build changes from #10102 here and leaves the simple build changes there. I consider usage of LIBMULTIPROCESS_LIBS to be a simple build change. I could add some usages of LIBMULTIPROCESS_LIBS here, but then the complaint could be that I'm adding unneeded dependencies. You have to draw the line somewhere when you are splitting things up, and this seemed like a reasonable place.

hebasto commented:
re-ACK b919efadff3d0393f4a8c3c1dc735f7ac5c665bb, rebased only since my [previous](https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-605514556) review.

theuni commented:
Uh, was this actually ready to go in? Have we determined for sure that this is the approach we want? Has libmultiprocess been well-reviewed?

Maybe I missed the discussion, but I thought this was still at the RFC stage.

ryanofsky commented:
> Uh, was this actually ready to go in? Have we determined for sure that this is the approach we want? Has libmultiprocess been well-reviewed?

Definitely not reviewed enough to be turned on. This PR is just adding options to build with it which are off by default and marked experimental.

MarcoFalke commented:
> Have we determined for sure that this is the approach we want?

@theuni No, we haven't. Multiprocess is experimental and turned off by default. We have 6 months to come up with a better solution or revert it before the 0.21.0 release.

theuni commented:
@MarcoFalke Well it requires c++17, so it won't fit into 0.21.0 with the schedule we've come up with iirc.

MarcoFalke commented:
0.21.0 releases are planned to be built with C++17, and the multiprocess build is disabled by default. Not to mention that this doesn't include a user-visible feature, so there is nothing to ship in either case.

The changes here add a build skeleton for the eventual multiprocess goal and some ci files to have it run on travis.

Some more context from IRC for archival purposes:

```
[18:40] <MarcoFalke> cfields: Are you assuming that multiprocess was merged for 0.20.0rc1?
[18:40] <cfields> MarcoFalke: no. It's an enormous change even for master.
[18:40] <fanquake> No. We are just confused why it was merged at all
[18:40] <cfields> afaik we've never merged a half-decided feature?
[18:41] <fanquake> Regardless of the questions Cory posted. Having a single ACK for what is the basis of a major architectural change for a project seems a bit weak.
[18:41] <MarcoFalke> I have reviewed the changes as well (not libmultiprocess)
[18:42] <MarcoFalke> If it is controversial, it should be reverted 
[18:42] <cfields> to be clear: I don't have a problem with this way forward afaik. I just think it deserves way more discussion.
[18:42] <cfields> Rather, I don't know if I have a problem. I assume anything ryanofsky comes up with is reasonable :)
[18:43] <MarcoFalke> It was not clear from the discussion in the pull request that this was not ready
[18:44] <cfields> MarcoFalke: I don't see how it can be ready unless the overall approach is agreed upon.
[18:45] <MarcoFalke> multiprocess has been discussed for three years as an idea to split the gui from the node (or even longer). I must have missed the discussions that multiprocess is unwanted.
[18:47] <MarcoFalke> Or controversial
[18:47] <fanquake> As Cory has just mentioned, we can't do it in 0.21.0 anyways, if it requires c++17 ?
[18:48] <MarcoFalke> Ok, so the goal is to revert and then schedule for 0.22, together with C++17?
[18:48] <fanquake> Given the discussion in #16684
[18:48] <gribble> https://github.com/bitcoin/bitcoin/issues/16684 | Discussion: upgrading to C++17 · Issue #16684 · bitcoin/bitcoin · GitHub
[18:48] <cfields> MarcoFalke: again, I'm not arguing against the idea. I just don't think the the specifics have been discussed.
[18:49] <MarcoFalke> Though, it doesn't require C++17, if the user doesn't ask for it
[18:50] <fanquake> MarcoFalke: can you clarify https://github.com/chaincodelabs/libmultiprocess/blob/master/CMakeLists.txt#L84 ?
[18:50] <MarcoFalke> fanquake: If you opt out of multiprocess, which is the default, you can go with C++11
[18:51] <cfields> So we've merged something we can't build?
[18:51] <fanquake> Sure, but that means we can't actually use multiprocess until after c++17 is a requirement for the project.
[18:51] <MarcoFalke> cfields: Gitian could use it
[18:52] <MarcoFalke> Aren't we building releases for 0.21.0 with C++17?
[18:52] <MarcoFalke> The project stays at C++11
[18:53] <MarcoFalke> Ok, to sum up: People like multiprocess as a concept, but the current implementation was not agree upon?
[18:53] <MarcoFalke> *agreed
[18:53] <ryanofsky> cfields, just to make sure there's no confusion, you know the whole multiprocess feature is implemented to be an optional add-on?
[18:54] <ryanofsky> it's not like there's a switch from single process to multiprocess, it's an option to build a new set of binaries alongside existing ones
[18:57] <MarcoFalke> ryanofsky: I think the disagreement is mostly about the additions to the depends/bitcoind Makefile and general concerns about libmultiprocess being ready
[18:58] <ryanofsky> right but this doesn't seem that different than the build change that added support for rapidcheck
[18:58] <ryanofsky> just an optional thing turned off by default that we had build/depends support for
[18:59] <cfields> ryanofsky: yes, but again, I thought that approach was at the RFC stage. When I reviewed it, I reviewed it on-merit, without arguing potential alternatives yet.
[19:00] <cfields> ryanofsky: I really really don't want to discourage this, that's not my aim. But I doubt I'm alone in being caught off-guard by the merge.
[19:00] <ryanofsky> sure, i'm just making sure there isn't some misunderstanding about the change
[19:02] <ryanofsky> whatever people think the next steps should be is fine with me
[19:07] <MarcoFalke> It is more invasive than rapidcheck, but to me it felt like the same concept: Merge it early when the code has received review, but mark it experimental so that it can be pulled out and reverted quickly if needed.
[19:09] <cfields> MarcoFalke: yeah, I see that now. To me, the merge implied a commitment to eventual default-ness because afaik that's the path we've always taken for things like this (thinking of univalue/secp256k1, at least).
[19:10] <fanquake> I think it would still be good for someone to outline how they expect this integration to play out. When does it get turned on by default? When do Cores build requirements change? When is review of the libmultiprocess lib/repo happening etc
[19:11] <cfields> ^^ +1
[19:11] <MarcoFalke> All of this must happen before it is taken out of experimental

ryanofsky commented:
> @MarcoFalke Well it requires c++17, so it won't fit into 0.21.0 with the schedule we've come up with iirc.

I'm little confused about this comment. What's the issue with c++17? Requirements to build bitcoin aren't changed by this PR. This PR adds a new `--enable-multiprocess` configure option and new `MULTIPROCESS=1` depends option. Only these new options which didn't exist before require c++17, so this is fully backwards compatible. Also the switch to c++17 from c++14 in libmultiprocess was done very recently in https://github.com/chaincodelabs/libmultiprocess/pull/25 to be able to use std::optional. It would be possible to have c++11 support if neccessary, too

MarcoFalke commented:
> It would be possible to have c++11 support if neccessary, too

C++17 will most likely be in Bitcoin Core before multiprocess is enabled by default, so I don't think effort should be put into C++11 support.

MarcoFalke commented:
Further discussion about reverting the changes now:


```
[19:18] <MarcoFalke> Is there anything to do before next weeks meeting? Right now it would be a clean git revert and I think everyone would be ok with reverting it until at least next week's meeting.
[19:19] <cfields> MarcoFalke: up to you. I feel heard either way :)
[19:20] <ryanofsky> either way's fine with me too
[19:20] <MarcoFalke> fanquake: Mind flipping a coin?
[19:21] <fanquake> I don't mind leaving it in, looking forward to thrashing all this out in some discussion

jnewbery commented:
> Some more context from IRC for archival purposes:

> Further discussion about reverting the changes now:

Where is all this discussion taking place? I can't find it in my local #bitcoin-core-dev log, and I don't see it in http://www.erisian.com.au/bitcoin-core-dev/ either

fanquake commented:
> Where is all this discussion taking place?

@jnewbery It's happening in the `#bitcoin-builds` channel. Previous logs available here: http://gnusha.org/bitcoin-builds/.

ryanofsky commented:
This PR, which was reverted in #18588, is a [proposed meeting topic](http://www.erisian.com.au/bitcoin-core-dev/log-2020-04-10.html#l-577). Background in https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Process-Separation

Topic could be broad, but two questions I'd like to get input on are:

1. What should steps and requirements be for moving [chaincodelabs/libmultiprocess](https://github.com/chaincodelabs/libmultiprocess) to [bitcoin-core/](https://github.com/bitcoin-core/), assuming this would be a better home for the library.

2. What should review requirements be for this PR, which adds `--enable-multiprocess` autoconf option (disabled by default and labeled experimental), and a travis build, and a depends definition to build [libmultiprocess](https://github.com/chaincodelabs/libmultiprocess) (also disabled by default and labeled experimental)?
